File: backend/internal/domain/poem.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

type PoemFormat string

const (
	FormatFreeVerse          PoemFormat = "free_verse"
	FormatHaiku              PoemFormat = "haiku"
	FormatSonnet             PoemFormat = "sonnet"
	FormatLimerick           PoemFormat = "limerick"
	FormatIambicPentameter   PoemFormat = "iambic_pentameter"
	FormatRhymingCouplets    PoemFormat = "rhyming_couplets"
	FormatCustom             PoemFormat = "custom"
)

type ApprovalMode string

const (
	ApprovalOpen     ApprovalMode = "open"
	ApprovalRequired ApprovalMode = "approval_required"
	ApprovalClosed   ApprovalMode = "closed"
)

type Poem struct {
	ID              uuid.UUID    `json:"id"`
	AuthorID        uuid.UUID    `json:"authorId"`
	Author          *User        `json:"author,omitempty"`
	Title           string       `json:"title"`
	Description     string       `json:"description"`
	Format          PoemFormat   `json:"format"`
	FormatRulesJSON string       `json:"formatRules,omitempty"`
	ApprovalMode    ApprovalMode `json:"approvalMode"`
	MaxStanzas      *int         `json:"maxStanzas"`
	IsHallOfFame    bool         `json:"isHallOfFame"`
	LikeCount       int          `json:"likeCount"`
	StanzaCount     int          `json:"stanzaCount"`
	CommentCount    int          `json:"commentCount"`
	Stanzas         []Stanza     `json:"stanzas,omitempty"`
	CreatedAt       time.Time    `json:"createdAt"`
	UpdatedAt       time.Time    `json:"updatedAt"`
}

type StanzaStatus string

const (
	StanzaApproved StanzaStatus = "approved"
	StanzaPending  StanzaStatus = "pending"
	StanzaRejected StanzaStatus = "rejected"
)

type Stanza struct {
	ID             uuid.UUID    `json:"id"`
	PoemID         uuid.UUID    `json:"poemId"`
	AuthorID       uuid.UUID    `json:"authorId"`
	Author         *User        `json:"author,omitempty"`
	Text           string       `json:"text"`
	Position       int          `json:"position"`
	LiteraryDevice string       `json:"literaryDevice,omitempty"`
	Status         StanzaStatus `json:"status"`
	CreatedAt      time.Time    `json:"createdAt"`
}

File: frontend/src/types/poem.ts
import type { User } from './user';

export type PoemFormat =
  | 'free_verse'
  | 'haiku'
  | 'sonnet'
  | 'limerick'
  | 'iambic_pentameter'
  | 'rhyming_couplets'
  | 'custom';

export type ApprovalMode = 'open' | 'approval_required' | 'closed';

export type StanzaStatus = 'approved' | 'pending' | 'rejected';

export interface Poem {
  id: string;
  authorId: string;
  author?: User;
  title: string;
  description: string;
  format: PoemFormat;
  formatRules?: string;
  approvalMode: ApprovalMode;
  maxStanzas?: number;
  isHallOfFame: boolean;
  likeCount: number;
  stanzaCount: number;
  commentCount: number;
  stanzas?: Stanza[];
  createdAt: string;
  updatedAt: string;
}

export interface Stanza {
  id: string;
  poemId: string;
  authorId: string;
  author?: User;
  text: string;
  position: number;
  literaryDevice?: string;
  status: StanzaStatus;
  createdAt: string;
}

export interface CreatePoemInput {
  title: string;
  description?: string;
  text: string;
  format: PoemFormat;
  approvalMode: ApprovalMode;
  maxStanzas?: number;
}

export interface SubmitStanzaInput {
  text: string;
  literaryDevice?: string;
}
